ethash/src/progpow.rs

```diff
@@ -0,0 +1,636 @@
+	match r % 4 {
+		0 => a.wrapping_mul(33).wrapping_add(b),
+		1 => (a ^ b).wrapping_mul(33),
+		2 => a.rotate_left(((r >> 16) % 31) + 1) ^ b,
+		_ => a.rotate_right(((r >> 16) % 31) + 1) ^ b,
+	}
+}
+
+fn math(a: u32, b: u32, r: u32) -> u32 {
+	match r % 11 {
+		0 => a.wrapping_add(b),
+		1 => a.wrapping_mul(b),
+		2 => ((a as u64).wrapping_mul(b as u64) >> 32) as u32,
+		3 => a.min(b),
+		4 => a.rotate_left(b),
+		5 => a.rotate_right(b),
+		6 => a & b,
+		7 => a | b,
+		8 => a ^ b,
+		9 => a.leading_zeros() + b.leading_zeros(),
+		_ => a.count_ones() + b.count_ones(),
+	}
+}
+
+fn progpow_init(seed: u64) -> (Kiss99, [u32; PROGPOW_REGS], [u32; PROGPOW_REGS]) {
+	let z = fnv1a_hash(FNV_HASH, seed as u32);
+	let w = fnv1a_hash(z, (seed >> 32) as u32);
+	let jsr = fnv1a_hash(w, seed as u32);
+	let jcong = fnv1a_hash(jsr, (seed >> 32) as u32);
+
+	let mut rnd = Kiss99::new(z, w, jsr, jcong);
+
+	// Create a random sequence of mix destinations for merge() and mix sources
+	// for cache reads guarantees every destination merged once and guarantees
+	// no duplicate cache reads, which could be optimized away. Uses
+	// Fisher-Yates shuffle.
+	let mut mix_seq_dst = [0u32; PROGPOW_REGS];
+	let mut mix_seq_cache = [0u32; PROGPOW_REGS];
+	for i in 0..mix_seq_dst.len() {
+		mix_seq_dst[i] = i as u32;
+		mix_seq_cache[i] = i as u32;
+	}
+
+	for i in (1..mix_seq_dst.len()).rev() {
+		unsafe {
+			// NOTE: `i` takes values from the range [1..PROGPOW_REGS] and `j`
+			// takes values from the the range [0..i]. This way it is guaranteed
+			// that the indices are always within the range of `mix_seq_dst` and
+			// `mix_seq_cache` and we can skip the bounds checking.
+			let j = rnd.next_u32() as usize % (i + 1);
+			::std::ptr::swap(&mut mix_seq_dst[i], mix_seq_dst.get_unchecked_mut(j));
+
+			let j = rnd.next_u32() as usize % (i + 1);
+			::std::ptr::swap(&mut mix_seq_cache[i], mix_seq_cache.get_unchecked_mut(j));
+		}
+	}
+
+	(rnd, mix_seq_dst, mix_seq_cache)
+}
+
+pub type CDag = [u32; PROGPOW_CACHE_WORDS];
+
+fn progpow_loop(
+	seed: u64,
+	loop_: usize,
+	mix: &mut [[u32; PROGPOW_REGS]; PROGPOW_LANES],
+	cache: &[Node],
+	c_dag: &CDag,
+	data_size: usize,
+) {
+	// All lanes share a base address for the global load. Global offset uses
+	// mix[0] to guarantee it depends on the load result.
+	let g_offset = mix[loop_ % PROGPOW_LANES][0] as usize %
+		(64 * data_size / (PROGPOW_LANES * PROGPOW_DAG_LOADS));
+
+	// 256 bytes of dag data
+	let mut dag_item = [0u32; 64];
+
+	// Fetch DAG nodes (64 bytes each)
+	for l in 0..PROGPOW_DAG_LOADS {
+		let index = g_offset * PROGPOW_LANES * PROGPOW_DAG_LOADS + l * 16;
+		let node = calculate_dag_item(index as u32 / 16, cache);
+		dag_item[l * 16..(l + 1) * 16].clone_from_slice(node.as_words());
+	}
+
+	// Lanes can execute in parallel and will be convergent
+	for l in 0..mix.len() {
+		// Initialize the seed and mix destination sequence
+		let (mut rnd, mix_seq_dst, mix_seq_cache) = progpow_init(seed);
```

Comment:
Here I think it would be better if this line was out of the loop and we just clone `rnd` on every iteration (kiss need derive clone).

Reply:
Good catch!